src/main/java/uk/gov/nationalarchives/pdi/step/jena/model/JenaModelStep.java

```diff
@@ -320,20 +322,26 @@ private Object convertSqlValueToRdf(final Object sqlValue, @Nullable final RDFDa
             } else if (sqlValue instanceof Number) {
                 return ((Number)sqlValue).toString();
 
-            } else {
-                // fallback
-                logBasic("convertSqlValueToRdfLiteralValue: required xsd:string but was given: {0}, unsure how to convert... Will default to Object#toString()!", sqlValue.getClass());
-                return sqlValue.toString();
             }
 
-        } else if (rdfDatatype.getURI().equals("http://www.w3.org/2001/XMLSchema#dateTime")) {
+        } else if (rdfDatatype.getURI().equals(XSD_NAMESPACE_IRI + "dateTime")) {
             // to xsd:dateTime
             if (sqlValue instanceof String) {
                 return sqlValue;
 
             } else if (sqlValue instanceof java.sql.Date || sqlValue instanceof java.sql.Timestamp || sqlValue instanceof java.util.Date) {
                 return (java.util.Date) sqlValue;
             }
+
+        } else if (rdfDatatype.getURI().equals(RDF_NAMESPACE_IRI + "XMLLiteral")) {
+            // to rdf:XMLLiteral
+            if (sqlValue instanceof String) {
+                return sqlValue;
+
+            } else if (sqlValue instanceof byte[]) {
```

Comment:
This doesn't belong here in my opinion.
The input field type (byte array here) has nothing to do with the output datatype (XML literal here).
Input is byte array if Pentaho didn't convert it yet, not if output is XML.
Decoding input from byte array to whatever it represents should happen outside of output datatype choice and encoding.

For exampe there could be an input flowing from a previous step that is actually a `java.sql.Date`, but Pentaho decides to only convert it lazily (i.e. not convert it), in which case it will also be a byte array.

I think such conversion should happen outside of the plugin. It should be implemented instead as part of the transformation, either by turning lazy conversion off (if all fields are used and there's no performance penalty because conversion would have to happen anyway somewhere) or by using the [_Select Values_ step](https://help.pentaho.com/Documentation/8.2/Products/Data_Integration/Transformation_Step_Reference/Select_Values).

Reply:
I am not sure I follow you... Pentaho does not necessarily know how to convert our types, the input type to output type mapping can only be managed by ourselves.

Converting a `byte[]` to a `Date` (e.g. `xs:date`) is not the same as converting a `byte[]` to a `String` (e.g. `xsd:string`), in the same way that converting a `String` to an `xsd:date`, is not the same as converting a `String` to an `xsd:string`. If anything this conversion method needs to be expanded, so far I have only added what we have needed as we have gone along.

if you still disagree then it is probably best to discuss on a quick call.